src/main/java/com/google/sps/servlets/BooksServlet.java

```diff
@@ -0,0 +1,151 @@
+package com.google.sps.servlets;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+import com.google.gson.Gson;
+import javax.servlet.annotation.WebServlet;
+import javax.servlet.http.HttpServlet;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import com.google.sps.data.Book;
+import java.net.URL;
+import java.net.HttpURLConnection;
+import java.util.Scanner;
+import com.google.gson.JsonArray;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.google.gson.JsonElement;
+import java.io.UnsupportedEncodingException;
+import java.nio.charset.StandardCharsets;
+import java.net.URLEncoder;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+
+/**
+ * Servlet for generating books from keywords.
+ */
+@WebServlet("/books")
+public class BooksServlet extends HttpServlet {
+
+  private final int NUM_BOOKS_PER_KEYWORD = 3;
+  private List<String> keywords = new ArrayList<String>(
+    List.of("Black Lives Matter", "COVID-19")
+  );
+  private LinkedHashMap<String, List<Book>> booksMap = new LinkedHashMap<>();
+
+  @Override
+  public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
+    String encodedKeyword;
+    for (String keyword : keywords) {
+      encodedKeyword = encodeTerm(keyword);
+      addBooksForTerm(keyword, encodedKeyword);
+    }
+    String json = makeBooksJson();
+    response.getWriter().println(json);
+  }
+
+ /**
+  * Encodes the given term so that it can fit in a url.
+  * @return An encoded string
+  */
+  private String encodeTerm(String term) {
+    try {
+      return URLEncoder.encode(term, StandardCharsets.UTF_8.toString());
+    } catch (UnsupportedEncodingException ex) {
+      throw new RuntimeException(ex.getCause());
+    }
+  }
+
+  private String makeBooksJson() {
+    Gson gson = new Gson();
+    Map<String, Object> books = new HashMap<>();
+    books.put("books", booksMap);
+    String json = gson.toJson(books); 
+    return json;
+  }
+
+ /**
+  * Adds Book object to books from JSON results of Guardian query.
+  */
+  private void jsonToBooks(String jsonString, String keyTerm) {
+    JsonObject responseObject = new JsonParser().parse(jsonString).getAsJsonObject();
+    // Check to see if json has enough entries.
+    int numResults = responseObject.get("totalItems").getAsInt();
+    int resultsToShow = NUM_BOOKS_PER_KEYWORD;
+    if (numResults < resultsToShow) {
+      if (numResults > 0) {
+        resultsToShow = numResults;
+      } else {
+        System.err.println("Error: no results found for query");
+        return;
+      }
+    }
+    JsonArray booksJsonArray = responseObject.getAsJsonArray("items");
+    List<Book> books = new ArrayList<Book>();
```

Comment:
[Nit] You do not need to define the generic type for the constructor.
`List<Book> books = new ArrayList<>();`